### PR TITLE
Feature/iglistkit implementation

### PR DESCRIPTION
**Pull request by polikhill**

### CODE REVIEW

File: ReduxTestTask/Base/BaseClasses/BaseCollectionViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,21 @@
+//
+//  BaseCollectionViewCell.swift
+//  ZablotskyDental
+//
+//  Created by Dima Ripa on 8/27/18.
```
Comment (aLHF) [inline ReduxTestTask/Base/BaseClasses/BaseCollectionViewCell.swift:5]: 🙂

File: ReduxTestTask/Classes/NewsList/View/NewsParentController.swift [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            .disposed(by: disposedBag)
+        
+        outputs.stateChanged
+        .subscribe()
+        .disposed(by: disposedBag)
+    }
+    
+    private func render(props: Props) {
+        if renderedProps?.items != props.items {
+            data = props.items.compactMap({ DiffableBox(value: $0, identifier: $0.diffIdentifier as NSObjectProtocol, equal: ==) })
```
Comment (aLHF) [inline ReduxTestTask/Classes/NewsList/View/NewsParentController.swift]: DiffableBox(...) should be created in makeProps function.

File: ReduxTestTask/Classes/NewsList/View/NewsParentController.swift [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+final class NewsParentController: UIViewController {
+
+    struct Props {
+        let items: [NewsCell.Props]
+        let error: String?
+        let isLoading: Bool
+    }
+    
+    private lazy var contentView = NewsListView()
+    private var data = [ListDiffable]()
```
Comment (aLHF) [inline ReduxTestTask/Classes/NewsList/View/NewsParentController.swift]: This data should be in the view. 

Controller and view have different props. Like this:

Props {
  let viewProps: NewsListView.Props
  let error: String?
  let isLoading: Bool
}

File: ReduxTestTask/Classes/NewsList/ViewModel/Middleware/NewsListFetchMiddleware.swift [hunk 4]
```diff
@@ -93,6 +93,7 @@ extension NewsList {
             Observable
                 .just(article)
                 .map ({ article -> Action in
+                    showArticle.onNext(article)
```
Comment (aLHF) [inline ReduxTestTask/Classes/NewsList/ViewModel/Middleware/NewsListFetchMiddleware.swift]: map should be used for transformations, not for side effects. Same for flatMap.

Use do(onNext:) for side effects

File: ReduxTestTask/Classes/NewsList/ViewModel/NewsListStore.swift [hunk 5]
```diff
@@ -67,8 +66,8 @@ extension NewsList {
         case is SelectedCell:
             break
             
-        case let action as ShowArticle:
-            newState.selectedArticle = action.article
+        case is ShowArticle:
```
Comment (aLHF) [inline ReduxTestTask/Classes/NewsList/ViewModel/NewsListStore.swift]: let's remove this if we don't do any work here. We have the default case for this. Same for isSelectedCell

File: ReduxTestTask/Classes/NewsList/ViewModel/NewsListViewModel.swift [hunk 6]
```diff
@@ -55,16 +56,29 @@ extension NewsList {
...
             
-            let actionCreator = ActionCreator(inputs: inputs, newsService: service)
+            let actionCreator = ActionCreator(inputs: inputs)
             
             let stateChanges = actionCreator.actions
                 .do(onNext: store.dispatch)
                 .voidValues()
+    
+        
+//            let article = store.state
```
Comment (aLHF) [inline ReduxTestTask/Classes/NewsList/ViewModel/NewsListViewModel.swift]: let's remove commented code

File: ReduxTestTask/Classes/NewsList/View/NewsListView.swift [hunk 7]
```diff
@@ -9,50 +9,80 @@
...
+        loader.isHidden = true
+        loader.hidesWhenStopped = true
+        addSubview(loader)
+        loader.center = center
     }
     
     private func toggleLoading(on: Bool) {
         if on {
-            refreshControl.beginRefreshing()
+            DispatchQueue.main.async {
```
Comment (aLHF) [inline ReduxTestTask/Classes/NewsList/View/NewsListView.swift]: render should be called on UI Thread so we don't need to dispatch here

File: ReduxTestTask/Classes/NewsList/ViewModel/NewsListProps.swift [hunk 8]
```diff
@@ -14,17 +14,18 @@ extension NewsList {
         return NewsListViewController.Props(
             contentViewProps: NewsListView.ContentViewProps(
                 isLoading: state.isLoading,
-                items: state.fetchedNews.compactMap { makeNewsCellProps(from: $0) }),
+                items: state.fetchedNews.compactMap { makeNewsCellDiffableProps(from: $0) }),
```
Comment (aLHF) [inline ReduxTestTask/Classes/NewsList/ViewModel/NewsListProps.swift]: why compactMap?

File: ReduxTestTask/Classes/NewsList/View/NewsListController.swift [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        return diffableNews
+    }
+    
+    func sectionController(
+        _ sectionController: ListBindingSectionController<ListDiffable>,
+        cellForViewModel viewModel: Any, at index: Int
+        ) -> UICollectionViewCell & ListBindable {
+
+        
+        guard let cell = collectionContext?.dequeueReusableCell(withNibName: "\(NewsCell.self)", bundle: nil, for: self, at: index) as? NewsCell else { fatalError() }
```
Comment (aLHF) [inline ReduxTestTask/Classes/NewsList/View/NewsListController.swift]: you have an extension for tableView or collectionView with a helper for dequeue. Let's write one for IGListKit too

File: ReduxTestTask/Classes/NewsList/ViewModel/NewsListProps.swift (hunk 8 above)
Comment (polikhill) [reply ReduxTestTask/Classes/NewsList/ViewModel/NewsListProps.swift]: I forgot that compactMap for optional init

**Pull request merged**